objects/src/accounts/account_id.rs

```diff
@@ -134,336 +129,471 @@ impl FromStr for AccountStorageMode {
     }
 }
 
+// ACCOUNT VERSION
+// ================================================================================================
+
+/// The version of an [`AccountId`].
+///
+/// Each version has a public associated constant, e.g. [`AccountVersion::VERSION_0`].
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub struct AccountVersion(u8);
+
+impl AccountVersion {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    const VERSION_0_NUMBER: u8 = 0;
+    /// Version 0 of the [`AccountId`].
+    pub const VERSION_0: AccountVersion = AccountVersion(Self::VERSION_0_NUMBER);
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the version number.
+    pub const fn version_num(&self) -> u8 {
+        self.0
+    }
+}
+
 // ACCOUNT ID
 // ================================================================================================
 
-/// Unique identifier of an account.
+/// The identifier of an [`Account`](crate::accounts::Account).
+///
+/// # Layout
 ///
-/// Account ID consists of 1 field element (~64 bits). The most significant bits in the id are used
-/// to encode the account' storage and type.
+/// An `AccountId` consists of two field elements and is layed out as follows:
 ///
-/// The top two bits are used to encode the storage type. The values [PRIVATE] and [PUBLIC]
-/// encode the account's storage type. The next two bits encode the account type. The values
-/// [FUNGIBLE_FAUCET], [NON_FUNGIBLE_FAUCET], [REGULAR_ACCOUNT_IMMUTABLE_CODE], and
-/// [REGULAR_ACCOUNT_UPDATABLE_CODE] encode the account's type.
+/// ```text
+/// 1st felt: [zero bit | random (55 bits) | storage mode (2 bits) | type (2 bits) | version (4 bits)]
+/// 2nd felt: [anchor_epoch (16 bits) | random (40 bits) | 8 zero bits]
```

Comment:
The only reason for the zero bit was to ensure felt validity as far as I remember. Since the metadata in the first felt is part of the hash output, I think we don't need to enforce any constraints, actually, because the hash function will always produce a valid felt in the first place. Not entirely sure now if I'm missing something.

We only had this general requirement of "at least one zero in the upper 32 bits" for the second felt because other data may be embedded in the second felt when included in other layouts. However, for the first felt we don't have that requirement because it is always included as-is in other layouts without being modified.

Reply:
Ah yes - since we are always doing PoW to get the right "shape" of the first felt, we don't need to do anything else with. So, we should be able to get rid of requiring the first bit to be zero without introducing any other restrictions.